Userland/Applications/PartitionEditor/PartitionModel.cpp

```diff
@@ -66,27 +66,32 @@ GUI::Variant PartitionModel::data(GUI::ModelIndex const& index, GUI::ModelRole r
     return {};
 }
 
+static ErrorOr<NonnullOwnPtr<Core::File>> clone_device_file_same_fd(NonnullOwnPtr<Core::File>& device)
+{
+    return Core::File::adopt_fd(device->fd(), Core::File::OpenMode::Read, Core::File::ShouldCloseFileDescriptor::No);
+}
```

Comment:
I wonder if we could possibly do some `MaybeOwned` trickery instead of this.

Something like handing out non-owning `MaybeOwned<Core::File>` to the `try_to_initialize` functions, and then hand off the actual owning `MaybeOwned` later once we have determined which type of partition table we have.

```cpp
auto file = TRY(Core::File::open(path, Core::File::OpenMode::Read));

auto mbr_table_or_error = Partition::MBRPartitionTable::try_to_initialize(MaybeOwned<Core::File>(*file));
if (!mbr_table_or_error.is_error()) {
    dbgln("Found MBR partition table on {}", path);
    m_partition_table = move(mbr_table_or_error.value());
    m_partition_table.adopt_backing_device(move(file));
    invalidate();
    return {};
}
```

Not sure if everyone would consider that strictly better though.

Reply:
I went with the solution that the caller either needs to know in advance which type of table they expect, or they have to keep an OwnPtr<File> around to keep the fd alive.

This way, we definitely don't leak fds, and if anything goes wrong then the fd is always immediately closed (and therefore the programmer should notice).